### DISCUSSION TITLE
How to incorporate many coordination information in a Transform4f variable

### DISCUSSION THREAD

**Discussion by yangocean-sudo** [mitsuba-renderer/mitsuba3#614] [Q&A] [answered]
Hello,

We are trying to simulate the photon as one light ray in mitsuba. The simulation will contain millions of light rays, and after the discussion, using one single emitter is a better choice than using millions of emitters. In the emitter we used before, we hardcore the sample_ray function in spot.cpp, which fixed the local_dir to one value: `(0.f, 0.f, 1.f)`. I noticed that our emitter is very similar to the spotlight emitter: our emitter is to emit rays from the different original points in different directions; while the spotlight emitter is to emit rays from one point to different directions based on the cone. 
For now, the emitter reads the binary file which contains the photons' information, and converts them to Transform4f variables. My idea is to pass a list that contains the Transform4f elements to the sample_ray function. In spot.cpp, the sample_ray function returns ` Ray3f(si.p, m_to_world.value() * local_dir, time, wavelengths), depolarizer<Spectrum>(spec_weight * falloff / pdf_dir)`, the parameter d is constructed by `m_to_world.value() * local_dir`, while local_dir is a Vector3f element and it is made up of many elements of triplets. It does make sense because it determines the direction of the lights.
My question is how to create a Transform4f or Vector4f(?) variable like local_dir, that contains all the information of photons so that I can have "photons_tranfo * local_dir" as the return ray's direction parameter, where the local_dir is a fixed vector 3f. Is it possible? 

Best wishes,
Yangyang

**Comment by njroussel** [comment, marked as answer]
Hi @yangocean-sudo 

If I understood correctly, this is how you could instantiate a vectorized `Transform4f` from a `std::vector` which we'll assume you filled from your binary file.

```cpp
if constexpr (dr::is_jit_v<Float>) {
    size_t N = 3;
    dr::Matrix<Float, 4> matrix = dr::zeros<dr::Matrix<Float, 4>>(N);

    std::vector<float> values{0, 1, 2};
    matrix.entry(0, 0) = dr::load<Float>(values.data(), N);
    matrix.entry(0, 1) = dr::load<Float>(values.data(), N);
    (...)

    Transform4f tf(matrix);
    std::cout << tf << std::endl;
}
```
(this snippet only works for JIT variants, making it work for scalar variants would require some more code)

Another important bit is that now `local_dir` should be a `ScalarVector3f` as it is no longer vectorized but rather a single vector.

**Reply by yangocean-sudo** [reply]
Hi @njroussel, thank you for your reply, it's very helpful. With your snippet, I decided to create a Transform4f variable in the `sampel_ray` function first and see what is the result, but I faced one error about incompatible size.
1. I firstly created a vectorized `Transform4f`, which contains coordination information about three photons, and it looks like this:
```
[[[-0.30703, 0, 0.951709, 0],
  [0.951709, 0, 0.307003, -1],
  [0, 1, 0, 0],
  [0, 0, 0, 1]],
 [[-0.325991, 0, 0.945373, 0],
  [0.945373, 0, 0.325991, -1],
  [0, 1, 0, 0],
  [0, 0, 0, 1]],
 [[-0.316228, 0, 0.948683, 0],
  [0.948683, 0, 0.316228, -1],
  [0, 1, 0, 0],
  [0, 0, 0, 1]]]
```
2. I changed the variable type of `local_dir` to `ScalarVector3f` :
`ScalarVector3f local_dir = ScalarVector3f(0.f, 0.f, 1.f);`
3. I return the ray and spectrum based on the Transform4f:
`  return { Ray3f(tf.translation(), tf * local_dir, time, wavelengths), depolarizer<Spectrum>(spec_weight * falloff / pdf_dir) };`
By the way, I set `falloff` to 1.0f and `pdf_dir` to 445029.
4. The ray seems normal and it looks like:
```
Ray3fDL[
  o = [[0, -1, 0],
       [0, -1, 0],
       [0, -1, 0]],
  d = [[0.951709, 0.307003, 0],
       [0.945373, 0.325991, 0],
       [0.948683, 0.316228, 0]],
  maxt = [3.40282e+38],
  time = [0],
]
```
5. When I run the code, the Jupyter Notebook displays this error information:
```
File ~/Documents/mitsuba3/build/python/mitsuba/python/util.py:518, in render(scene, params, sensor, integrator, seed, seed_grad, spp, spp_grad)
    514 elif seed_grad == seed:
    515     raise Exception('The primal and differential seed should be different '
    516                     'to ensure unbiased gradient computation!')
--> 518 return dr.custom(_RenderOp, scene, sensor, params, integrator,
    519                  (seed, seed_grad), (spp, spp_grad))

File ~/Documents/mitsuba3/build/python/drjit/router.py:5585, in custom(cls, *args, **kwargs)
   5583     output = inst.eval(*_dr.detach(kwargs['args']))
   5584 else:
-> 5585     output = inst.eval(**{ k: _dr.detach(v) for k, v in kwargs.items() })
   5586 if _dr.grad_enabled(output):
   5587     raise RuntimeError("drjit.custom(): the return value of CustomOp.eval() "
   5588                        "should not be attached to the AD graph!")

File ~/Documents/mitsuba3/build/python/mitsuba/python/util.py:373, in _RenderOp.eval(self, scene, sensor, params, integrator, seed, spp)
    370 self.spp = spp
    372 with dr.suspend_grad():
--> 373     return self.integrator.render(
    374         scene=self.scene,
    375         sensor=sensor,
    376         seed=seed[0],
    377         spp=spp[0],
    378         develop=True,
    379         evaluate=False
    380     )

RuntimeError: jit_var_vcall(): input/output arrays have incompatible size!
```
It seems the error happened in the `sample_emitter_ray` function, especially on the code ` std::tie(ray, weight) = emitter->sample_ray(time, sample_1_re, sample2, sample3, active);` in `scene.cpp` file. However, after comparing it with the ray3f which is generated by other emitters, I didn't find slight differences. 
It will be nice if you give me some hints, thank you for your time.

**Reply by njroussel** [reply]
This is expected. In your example, you have now "hardcoded" the output size to 3. However the input has a completely different size, namely the width of your wavefront (width * height * spp).

In mitsuba the entire rendering is vectorized over the rays - so there are typically `width * height * spp` lanes at any point in the code.

**Reply by njroussel** [reply]
As the method name suggests, you need to sample rays from your emitter. This is usually a random process. In addition, the number of rays needs to match the number of input samples.

In your case, you'll most likely need something like this:
```
using Matrix = dr::Matrix<Float, 4>;
UInt32 index =  dr::arange<UInt32>(10) % 3;
Matrix gathered = dr::gather<Matrix>(matrix, index);
std::cout << gathered << std::endl;
``` 
Where `index` should be a random vector of the appropriate width (replace 10 by whatever the input width is) and where `matrix` is your vectorized matrix of arbitrary size which you need to sample from.

**Reply by yangocean-sudo** [reply]
Thanks for the quick response, I've thought about it for a time and I'm not sure whether I understand everything.
 I'm confused about what the input width is, for my understanding, the input width is the width of the wavefront, which is the length of `wavelength_sample` in the sample_ray function. (Btw, is there a method to get the length of `Float` variable `wavelength_sample` directly?)
I don't get the idea that " index should be a random vector of the appropriate width ". If the value of the index is got by using the function `dr::arange`, it should look like `[0, 1, 2, 0, 1, ... , 0, 1, 2]`, but it is not a **random** vector.
If I use the length of `wavelength_sample` to replace the value `10`, the error is: 
`operands have incompatible sizes! (sizes: 3, 1048576)`
Iterating to make my matrix size from 3 to the length of `wavelength_sample`  may will solve this error, but I'm not sure if I need to do that because "matrix is the vectorized matrix of arbitrary size which need to sample from."

**Reply by njroussel** [reply]
`dr::width(var)` will give you the width of any variable. 

Sorry, my previous comment mislead you. The `dr::arange()` was just an example of an index of the appropriate size. You should rather compute it using the input samples. I didn't know if in your setup you wanted to importance sample some photons specifically, or just have a uniform sampling.

Indeed, you don't need to repeat your matrices. That's the point of the `gather` in my snippet. As we established above, the indices will have the same width as `wavelength_sample` whereas the `matrix` will have some completely other size. The `gather` operation is what allows you to bridge that gap. The gathered result will have the same width as you `index` hence the same width as your `wavelength_sample`.

**Reply by yangocean-sudo** [reply]
Thanks for your reply. 

I found one interesting thing that is the `sample_1_re` variable, which is in the function `sample_emitter_ray` in `scene.cpp` , has the width value `1048576 `, which is reasonable, because it is sampled in the integrator, I use `ptracer` in this case. However, once this variable is passed to `sample_ray` function in emitters (I tried both my custom emitter and the spot emitter) as the parameter `wavelength_sample`, the value of `dr::width(wavelength_sample)` is `1`. (By the way, is there a method to print out the variable such as `wavelength_sample` in DrJit's loop?) I don't understand how it changes since it should be passed from `scene.cpp` to the emitter file directly.

The other question is that in the `gather` function we discussed before, from my understanding, the variable `matrix` should be filled from my binary file, instead of full of zeros at that step. However, if I did so, it shows "Python quit unexpectedly" and the error message references a function called "jitc_fail" and "jitc_llvm_compile", indicating that there may have been an issue with the compilation process. On the other hand, if I create a Matrix with zeros, like `dr::Matrix<Float, 4> matrix = dr::zeros<dr::Matrix<Float, 4>>(N);`,  `gather` it with the vector index, and then set the value for the first three matrices from my binary file (say there are only three photons in this file), it will work.

After reading the document content about dr::gather(), I think the gather function will combine the index vector and the matrix in this way:
For example, I have one vector: `index = [0,1,2,0,1,2]`, and I have three matrices: `matrix1, matrix2, matrix3.`After using the `gather` function, the Matrix `gathered` will be:
`gathered = [matrix1[0] , matrix2[1], matrix3[2], matrix1[0] , matrix2[1], matrix3[2]]`
If I instantiate my matrix in my second way, the matrix `gathered` will be:
`gathered = [matrix1[0] , matrix2[1], matrix3[2], zeromatrix[0] , zeromatrix[1], zeromatrix[2]]` 
because I `gather` the matrix that is full of zeros with the vector first. 
The first approach makes more sense but it failed during the compilation process. 
Please let me know if I didn't explain it clearly or if I misunderstand something. Thank you for your time.

**Reply by njroussel** [reply]
This is an artifact of how recorder virtual function call works, I believe. You can disable recorded loops and recorded virtual function calls. This allows you to print/evaluate variables wherever you want which is practical for debugging. It does come with a significant performance and memory cost typically.
```
jit_set_flag(JitFlag::LoopRecord, false);
jit_set_flag(JitFlag::VCallRecord, false);
```

This might just be a typo in your example, but the gathered result should not have sub-indexing:
```
 gathered = [matrix1 , matrix2, matrix3, matrix1 , matrix2, matrix3]
```

I quickly tried this myself, and it seems to work. I can't tell what you're doing wrong without a clearer error message. I literally just put the two snippets I already sent together. Here's a diff: 
```diff
diff --git a/src/emitters/spot.cpp b/src/emitters/spot.cpp
index 0e1304bf..c9739e58 100644
--- a/src/emitters/spot.cpp
+++ b/src/emitters/spot.cpp
@@ -112,6 +112,24 @@ public:
         Assert(dr::all(m_cutoff_angle >= m_beam_width));
         m_uv_factor = dr::tan(m_cutoff_angle);

+        m_transforms = dr::zeros<dr::Matrix<Float, 4>>(3);
+        m_transforms.entry(0, 0) = dr::arange<Float>(0,  3);
+        m_transforms.entry(0, 1) = dr::arange<Float>(1,  4);
+        m_transforms.entry(0, 2) = dr::arange<Float>(2,  5);
+        m_transforms.entry(0, 3) = dr::arange<Float>(3,  6);
+        m_transforms.entry(1, 0) = dr::arange<Float>(4,  7);
+        m_transforms.entry(1, 1) = dr::arange<Float>(5,  8);
+        m_transforms.entry(1, 2) = dr::arange<Float>(6,  9);
+        m_transforms.entry(1, 3) = dr::arange<Float>(7,  10);
+        m_transforms.entry(2, 0) = dr::arange<Float>(8,  11);
+        m_transforms.entry(2, 1) = dr::arange<Float>(9,  12);
+        m_transforms.entry(2, 2) = dr::arange<Float>(10, 13);
+        m_transforms.entry(2, 3) = dr::arange<Float>(11, 14);
+        m_transforms.entry(3, 0) = dr::arange<Float>(12, 15);
+        m_transforms.entry(3, 1) = dr::arange<Float>(13, 16);
+        m_transforms.entry(3, 2) = dr::arange<Float>(14, 17);
+        m_transforms.entry(3, 3) = dr::arange<Float>(15, 18);
+
         dr::make_opaque(m_beam_width, m_cutoff_angle, m_uv_factor,
                         m_cos_beam_width, m_cos_cutoff_angle,
                         m_inv_transition_width);
@@ -157,8 +175,13 @@ public:
         MI_MASKED_FUNCTION(ProfilerPhase::EndpointSampleRay, active);

         // 1. Sample directional component
-        Vector3f local_dir = warp::square_to_uniform_cone(spatial_sample, (Float) m_cos_cutoff_angle);
-        Float pdf_dir = warp::square_to_uniform_cone_pdf(local_dir, (Float) m_cos_cutoff_angle);
+        ScalarVector3f local_dir(0, 0, 1);
+        UInt32 index = dr::floor(spatial_sample[0] * 3);
+
+        Matrix4f transforms = dr::gather<Matrix4f>(m_transforms, index);
+        Vector3f new_dir = Transform4f(transforms).transform_affine(local_dir);
+
+        Float pdf_dir = 1.f;

         // 2. Sample spectrum
         auto si = dr::zeros<SurfaceInteraction3f>();
@@ -170,7 +193,7 @@ public:

         Float falloff = falloff_curve(local_dir, active);

-        return { Ray3f(si.p, m_to_world.value() * local_dir, time, wavelengths),
+        return { Ray3f(si.p, new_dir, time, wavelengths),
                  depolarizer<Spectrum>(spec_weight * falloff / pdf_dir) };
     }

@@ -278,6 +301,7 @@ private:
     ref<Texture> m_texture;
     Float m_beam_width, m_cutoff_angle, m_uv_factor;
     Float m_cos_beam_width, m_cos_cutoff_angle, m_inv_transition_width;
+    Matrix4f m_transforms;
 };
```
Obviously this is not exactly what you want to do, the result is bogus. But the gather operation works as expected.

**Reply by yangocean-sudo** [reply]
Thank you for your reply. It's informative and helpful. With your help, now I can emit three light rays in one single emitter, which is very exciting. Thanks again.
I'm wondering if there is a way to make a Transform4f value do a `look_at` camera transformation. The value I put in the matrix `m_transforms` is a copy from `Transform4f trafo = Transform4f::look_at(origin, target, up);`, where the vector3f origin, target, up is made up by my binary input. I used camera coordination for the photons in the binary file because the data of photons are combined in that way and it makes more sense when I do some debugging. 

Based on your first reply, I'm wondering if I can add the `Transform4f` variable inside the `dr::Matrix matrix` instead of using `entry` function. If not, is there a way to convert `Float` type variable to `std::float`? What I am doing now is extracting the value from the `Transfrom4f` variable to the `std::vector<float> values` and then constructing the `matrix`, but I think there may exist a simpler approach.
Thank you for your time

> ```c++
> if constexpr (dr::is_jit_v<Float>) {
>     size_t N = 3;
>     dr::Matrix<Float, 4> matrix = dr::zeros<dr::Matrix<Float, 4>>(N);
> 
>     std::vector<float> values{0, 1, 2};
>     matrix.entry(0, 0) = dr::load<Float>(values.data(), N);
>     matrix.entry(0, 1) = dr::load<Float>(values.data(), N);
>     (...)
> 
>     Transform4f tf(matrix);
>     std::cout << tf << std::endl;
> }
> ```

**Reply by njroussel** [reply]
Yes, you can access a `Transform`'s matrix directly. It's [one of the struct's members](https://github.com/mitsuba-renderer/mitsuba3/blob/master/include/mitsuba/core/transform.h#L45).

**Reply by yangocean-sudo** [reply]
Sorry I haven't got the idea. I read the link and the drjit document about matrix for times but didn't find one way to modify the value (submatrix) of a Matrix4f variable. What I'm doing now is to get the `Transform4f` variables based on my binary file, and then create a `Matrix4f` variable, the size is equal to my photons' number. _I want to replace the submatrix to these `Transform4f` variables_, and then `gather` with the sample index, convert them to one `Transform4f` matrix, and finally get the direction vector.

Do I miss anything or which file do I also need to look at? Thank you for your time.

**Reply by njroussel** [reply]
I must still be misunderstanding what you're trying to achieve.

> modify the value (submatrix) of a Matrix4f variable. 

What do you mean by submatrix? The i-th matrix in the vectorized array of matrices?

**Reply by njroussel** [reply]
Here's some pseudo-code of what I think you should be doing in the constructor:
```
read std::vector<float> per coordinate for each origins, targets and ups from your file
load them each into separate Float variables with dr::load<Float>(std_vec_origins_x.data(), N)
create two Point3f and one Vector3f for origins, targets and ups from these Float variables
create Transform4f from these two Point3f and single Vector3f
get the matrix Matrix4f from the Transform
```
The `sample_ray()` can stay the same.

**Reply by yangocean-sudo** [reply]
Thank you very much! I found the variable type in Mitsuba is more potent than the standard one. With your pseudo-code and help, I finally created one emitter with a vectorized input.
